ADC_irq/Core/Src/main.c

```diff
@@ -8,101 +8,129 @@
 	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
 	{
 		ADC1->CR |= ADC_CR_ADSTP;			//if so, we have to set ADSTP bit
 		while(ADC1->CR & ADC_CR_ADSTP);		//and wait for it to clear
 	}
 	ADC1->CR |= ADC_CR_ADDIS;				//disables the ADC
 	while(ADC1->CR & ADC_CR_ADEN);
-	DISABLE_ALL_CHANNELS();					//and disables all channels
 
-	switch (Channel)
-	{
-		case CH_INT_TEMP:
-				ADC1->CHSELR |= ADC_CHSELR_CHSEL18;		//selecting the TSEN channel
-				ADC->CCR |= ADC_CCR_TSEN; 				//enables temperature sensor
-				Wait(TIME_10uSEC);						//we have to wait for the proper time for the Tsense to wake up
-		break;
-		case CH_V_REF:
-				ADC1->CHSELR |= ADC_CHSELR_CHSEL17;		//selecting the VREF channel
-				ADC->CCR |= ADC_CCR_VREFEN; 			//enables internal reference voltage
-		break;
-		default:
-				return (ADC_ERROR);	//ADC error
-		break;
-	}
+	ADC1->CHSELR = ADC_CHSELR_CHSEL17;		//selecting the VREF channel
+	ADC->CCR |= ADC_CCR_VREFEN; 			//enables internal reference voltage
 
-	ADC1->ISR|= ADC_ISR_ADRDY;				//clear the ADRDY bit by programming it to 1
 	ADC1->CR |= ADC_CR_ADEN;				//then we enable the ADC
 	while(!(ADC1->ISR & ADC_ISR_ADRDY));	//and wait for it to be ready. (Can be handled by interrupt)
 
+	/*									*
+	 * --- STARTING THE ADC READING --- *
+	 *									*/
+	ADC1->CR |= ADC_CR_ADSTART;				//starts the ADC
+
+	while(1)
+	{
+		if(flag_EOC)									//if THERE WAS an ADC interrupt
+		{
+			flag_EOC = 0;
+
+			switch (ADC1->CHSELR)
+			{
+				case ADC_CHSELR_CHSEL17:				//Internal reference voltage
+						v_ref = ADC1->DR;				//storing the data read in v_ref
+
+						//configuring the next channel to be read
+						ADC1->CHSELR = ADC_CHSELR_CHSEL18;
+						ADC->CCR 	|= ADC_CCR_TSEN; 	//enables temperature sensor
+						wait(TIME_10uSEC);				//we have to wait for the proper time for the Tsense to wake up
+						ADC1->CR 	|= ADC_CR_ADSTART;	//starts the ADC
+				break;
+				case ADC_CHSELR_CHSEL18:				//Internal temperature
+						temperature = ADC1->DR;			//we'll store the data read in temperature
+				break;
+				default:
+						//error management
+				break;
+			}
+
+			if(ADC1->CHSELR & ADC_CHSELR_CHSEL18)		//if we've finished reading all channels (could be done with a #define)
+			{											//we're going to calculate what we need
+				uint16_t aux = v_ref;					//we only use aux to calculate v_ref
+
+				//Vref calculation as in RM
+				v_ref = (3 * (int32_t) (*VREFINT_CAL));
+				v_ref = (v_ref * 1000) / aux;			//*1000 so we have Vref in mV
+														//the calculated internal reference voltage is now stored in v_ref
+
+				//temperature calculation as in RM example
+				temperature = ((temperature * VDD_APPLI / VDD_CALIB) - (int32_t) *TEMP30_CAL_ADDR);
+				temperature = temperature * (int32_t)(130 - 30);
+				temperature = temperature / (int32_t)(*TEMP130_CAL_ADDR - *TEMP30_CAL_ADDR);
+				temperature = temperature + 30;			//the calculated temperature is now stored in temperature
+

```

Comment:
I would rather see these temp and voltage calculations inside the switch case. What do you think?

Reply:
I know those are very simple examples, but I've removed them from the switch case thinking in a fast reading. We read channel 17, set channel 18, and read it, and then we calculate the values. Moving them to the switch case we would have read channel 17, calculate vref, set channel 18, read it, and calculate temp.

I think that, in this case, fast reading would be more appropriated than fast data recognition (data after calculation). In a case that data's true value would be more important, I would keep them in the switch case.